2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+```elixir
+defmodule PersonPresenter do
+  def trim_whitespace(name)
+    name
+    |> String.split()
+    |> Enum.join(" ")
+  end
+end
+```
+
+This is better. But recall that the data comes from a source you do not own. It's controlled by someone else and likely to change in unpredictable ways. Even code and data that you *do* own are susceptible to ever-evolving business requirements. The code you write today might seem concrete, but like Molly the dog, even seemingly concrete entities are not static. Code is a being in-process, constantly churning. Your source control repository's diff log is a more accurate mental model of your application than current files on disk.
+
+It's quite possible that people's names will someday arrive in a different form and that removing whitespace will no longer be the only operation required to prepare it for presentation. Naming a function after what it currently does leaves you no wiggle room for change. `trim_whitespace` explains the function's current implementation, but not what it means from the caller's perspective, nor what it means from an application domain perspective.
+
+Consider a new requirement. As it turns out, the data has not changed, but the product team would like names to be displayed in last name, first name order. The implementation will be simple enough to rework. But the name `trim_whitespace` has painted you into a corner.
+
+Should you rename the function, perhaps to `trim_whitespace_and_reorder`? Down that path lies madness. Naming a function after its implementation leads to expensive, hand-wringing decisions.
+
+But this simple example obscures an even more costly consideration. Your codebase depends on many messaging interfaces, as it should. In order to get anything done, your code must send messages somewhere. In other words, it must know the names of functions it can call. When your code calls a function, it becomes dependent upon the name of that function. Again, this is fine. It's better than fine; message interfaces provide the loosest form of coupling. Well-designed code depends on the messages it sends to direct collaborators and little else. Over time, many parts of your codebase may become dependent on a single message name.[^2]
+
+In your hypothetical app, a person's name is presented in several views. `trim_whitespace` is called from many places. Your codebase has become dependent upon that name. Alarm bells should be going off in your head when you see names representing implementation details spread throughout your codebase. Changing this function name at its source means changing it everywhere it's called. There's a code smell called Shotgun Surgery that occurs when making a single modification requires that you make changes in many other places. In the categories of code smells, Shotgun Surgery is considered a Change Preventer.
```

Comment:
Where does "Change Preventer" come from? Not another code smell? Your use of capitals suggests that it's a proper noun.

Reply:
It's probably a good idea to link to information on Shotgun Surgery and Change Preventer.

Reply:
Addressed in ac3fc886375a96bd557de03835275634397a4d63